src/zenml/steps/external_artifact.py

```diff
@@ -173,6 +178,32 @@ def upload_if_necessary(self) -> UUID:
             logger.info(
                 "Finished uploading external artifact %s.", artifact_id
             )
+        else:
+            response = None
+            if self._id:
+                response = Client().get_artifact(artifact_id=self._id)
+            elif self._pipeline_name and self._artifact_name:
+                pipeline = Client().get_pipeline(self._pipeline_name)
+                for artifact in pipeline.last_successful_run.artifacts:
+                    if artifact.name == self._artifact_name:
+                        response = artifact
+                        break
+
+            if response is None:
+                raise RuntimeError(
+                    f"Artifact with name `{self._artifact_name}` was not found "
+                    f"in last successful run of pipeline `{self._pipeline_name}`. "
+                    "Please check your inputs and try again."
+                )
+            elif response.artifact_store_id != artifact_store_id:
```

Comment:
I'd suggest to refactor this a bit since the `if response is None` does not apply to the ID case

```suggestion
            if self._id:
                response = Client().get_artifact(artifact_id=self._id)
            else:  # self._pipeline_name and self._artifact_name
                pipeline = Client().get_pipeline(self._pipeline_name)
                for artifact in pipeline.last_successful_run.artifacts:
                    if artifact.name == self._artifact_name:
                        response = artifact
                        break
                else:
                    raise RuntimeError(
                        f"Artifact with name `{self._artifact_name}` was not found "
                        f"in last successful run of pipeline `{self._pipeline_name}`. "
                        "Please check your inputs and try again."
                    )

            if response.artifact_store_id != artifact_store_id:
```

Reply:
mypy was going crazy with such an implementation, so I had to rewrite it like this.
https://github.com/zenml-io/zenml/actions/runs/5676919051/job/15384458918#step:5:33